### DISCUSSION TITLE
Collecting Database of Meshtastic Device's USB Serial Information

### DISCUSSION THREAD

**Discussion by broglep** [meshtastic/Discussions#109] [General]
Hi, I'm working on meshtastic home assistant integration and already have auto discovery for TCP (zeroconf) and bluetooth and would also like to add support for serial as well. Home assistant allows to auto discover based on [USB vendor and product ID](https://developers.home-assistant.io/docs/creating_integration_manifest/#usb) as well matching against manufacturer & description. Does anybody have an idea if there is already such a database or how to best build one?

I only have a small collection of devices and did look around a bit, seems like some devices use generic product / vendor ID from e.g. espressif, but have description that can be used to match, so in general it should be doable. Vendor and Product however will probably not be enough in all cases though.

I think the information from the most popular devices could be a good start as well.

**Comment by mariotti** [comment]
Just as curiosity, how could you match the TCP part? As in my case I have even difficulties to match a MAC address.

**Reply by broglep** [reply]
Do you mean how TCP auto discovery works with mDNS? It is currently not that robust, but it works and proper mDNS support is on its way for the firmware. Currently the integration looks for mDNS `_http._tcp.local.` and then checks that `Meshtastic` is the instance/host name (see [broglep/homeassistant-meshtastic/config_flow.py](https://github.com/broglep/homeassistant-meshtastic/blob/1a273433554b63a903b2d1d771f1cff50f1ff019/custom_components/meshtastic/config_flow.py#L284)). But now there is a proper `_meshtastic._tcp.local.` mDNS service being exposed (see https://github.com/meshtastic/firmware/pull/5355, https://github.com/meshtastic/Meshtastic-Android/pull/1401, https://github.com/meshtastic/firmware/pull/5503) that will properly identify a meshtastic node in a local network

**Comment by TerryFrench** [comment]
Some of my RAK boards only report the vid and pod of the serial finish chip that is on them over serial AFAIK. Same as would be seen from arduino ide. 

If through BLE there must be some way to fingerprint Meshtastic devices. Through usb or usb-serial you might have to poke for whatever is standard on Meshtastic flashed boards. Like trying to read Meshtastic settings maybe?

**Comment by broglep** [comment]
I ended up collecting hwids from https://github.com/meshtastic/firmware/tree/master/boards
It kind of should be fine for my use case, it will probably detect too many devices, so you end up some devices wrongfully offered by home assistant to be configured as meshtastic device. But as the user has the option to ignore devices or then actually confirm to configure them, this works fine. As the [home assistant integration](https://github.com/broglep/homeassistant-meshtastic) is still a custom component that needs to be installed manually, it won't affect unaware users.   

Actually opening a connection as part of the discovery and trying read out something settings to make sure that it is actually a meshtastic device felt a bit risky. The protocol requires you to send something first on the serial, so in worst case you end up writing some data to an unknown device that might trigger unexpected things on that device.
